WalletWasabi/WabiSabi/Crypto/CredentialIssuer.cs

```diff
@@ -196,31 +220,56 @@ public ICommitableCredentialsResponse PrepareResponse(CredentialsRequest registr
 			var areProofsValid = ProofSystem.Verify(transcript, statements, registrationRequest.Proofs);
 			if (!areProofsValid)
 			{
+				// Request was invalid, but all serial numbers were unused.
+				// Ensure nullifier set can't be clogged with invalid serial
+				// numbers. Valid serial numbers are only issued in response to
+				// valid requests, which in turn depend on valid ownership
+				// proofs and the banning mechanism, and they actually matter
+				// for double spending, so only they need to be stored in the
+				// nullifier set.
+				foreach (var serialNumber in presentedSerialNumbers)
+				{
+					SerialNumbers.Remove(serialNumber, out _);
+				}
+
 				throw new WabiSabiCryptoException(WabiSabiCryptoErrorCode.CoordinatorReceivedInvalidProofs);
 			}
 
-			// Issue credentials.
-			var credentials = requested.Select(x => IssueCredential(x.Ma, RandomNumberGenerator.GetScalar())).ToArray();
+			// After this point serial numbers are committed irrevocably, even
+			// if the request fails at a higher level, and the credentials
+			// were never revealed because `Commit` was not called. This is
+			// because any request that made it this far is formally valid, and
+			// thus any invalidity with regard to state is unambiguously a
+			// double spend attempt, and there is no point in allowing those
+			// serial numbers to be reused and the round to proceed.
+
+			// Issue the credentials, but they won't actually be returned until
+			// Commit() is called.
+			var credentials = requested.Select(x => IssueCredential(x.Ma, RandomNumberGenerator.GetScalar())).ToImmutableArray();
 
 			// Construct response.
 			var proofs = ProofSystem.Prove(transcript, credentials.Select(x => x.Knowledge), RandomNumberGenerator);
 			var macs = credentials.Select(x => x.Mac);
 			var response = new CredentialsResponse(macs, proofs);
 
-			var serialNumbers = presented.Select(x => x.S);
-			return new PreparedCredentialsResponse(this, response, registrationRequest.Delta, serialNumbers);
+			return new PreparedCredentialsResponse(this, response, registrationRequest.Delta);
 		}
 
-		private CredentialsResponse Commit(CredentialsResponse response, long delta, IEnumerable<GroupElement> serialNumbers)
+		private CredentialsResponse Commit(CredentialsResponse response, long delta)
 		{
-			// Register the serial numbers to prevent credential reuse.
-			foreach (var serialNumber in serialNumbers)
+			Interlocked.Add(ref _balance, delta);
```

Comment:
The balance remains negative.

Reply:
i don't understand... delta is signed?

Reply:
In previous version we added the balance only after validating the `balance + delta >=0` but here we add if even when it results in < 0. Is that okay?

Reply:
Oh, i see what you mean. The right way to handle this is for the coordinator to crash, and go offline, which we don't have a clear idea of how to do yet, since it indicates the entire round & crypto code is buggy, so there's no point in trying to gracefully recover from this error.